src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -1275,6 +1283,11 @@ ExecuteTransactionEnd(bool commit)
 	NodeConnectionEntry *participant;
 	bool completed = !commit; /* aborts are assumed completed */
 
+	if (xactParticipantHash == NULL)
```

Comment:
I guess this is here because we might be doing COPY, in which case we don't have a xactParticipantHash. That seems fine, though I wonder if it's worth distinguishing between ``XACT_MODIFICATION_DML`` and ``XACT_MODIFICATION_COPY`` to be explicit about different behaviour at transaction end. That would also allow showing a warning when trying to abort an already-completed COPY. 

I'm fine with leaving this the way it is for now, since we'll likely revisit the way COPY commits as part of connection management.